datafusion/datasource/src/projection.rs

```diff
@@ -0,0 +1,634 @@
+                                Arc::clone(&partition_literals[pci.in_partition_values]);
+                            return Ok(Transformed::yes(literal));
+                        }
+                    }
+                    Ok(Transformed::no(original_expr))
+                })
+                .data()
+                .expect("infallible transform");
+            ProjectionExpr::new(expr, projection.alias.clone())
+        })
+        .collect_vec();
+    ProjectionExprs::new(projections)
+}
+
+/// At a high level the goal of SplitProjection is to take a ProjectionExprs meant to be applied to the table schema
+/// and split that into:
+/// - The projection indices into the file schema (file_indices)
+/// - The projection indices into the partition values (partition_value_indices), which pre-compute both the index into the table schema
+///   and the index into the partition values array
+/// - A remapped projection that can be applied after the file projection is applied
+///   This remapped projection has the following properties:
+///     - Column indices referring to file columns are remapped to [0..file_indices.len())
+///     - Column indices referring to partition columns are remapped to [file_indices.len()..)
+///
+///   This allows the ProjectionOpener to easily identify which columns in the remapped projection
+///   refer to partition columns and substitute them with literals from the partition values.
+#[derive(Debug, Clone)]
+pub struct SplitProjection {
+    /// The original projection this [`SplitProjection`] was derived from
+    pub source: ProjectionExprs,
+    /// Column indices to read from file (public for file sources)
+    pub file_indices: Vec<usize>,
+    /// Pre-computed partition column mappings (internal, used by ProjectionOpener)
+    pub(crate) partition_columns: Vec<PartitionColumnIndex>,
+    /// The remapped projection (internal, used by ProjectionOpener)
+    pub(crate) remapped_projection: ProjectionExprs,
+}
+
+impl SplitProjection {
+    pub fn unprojected(table_schema: &TableSchema) -> Self {
+        let projection = ProjectionExprs::from_indices(
+            &(0..table_schema.table_schema().fields().len()).collect_vec(),
+            table_schema.table_schema(),
+        );
+        Self::new(table_schema.file_schema(), &projection)
+    }
+
+    /// Creates a new `SplitProjection` by splitting columns into file and partition columns.
+    ///
+    /// # Algorithm
+    /// Single-pass approach that combines extraction, classification, and remapping:
+    /// 1. Extract all unique column references from projection expressions
+    /// 2. Sort columns by original table index
+    /// 3. Classify each column as either file or partition based on file_schema length
+    /// 4. Assign final indices: file columns → [0..n), partition columns → [n..)
+    /// 5. Transform expressions once to remap all column references
+    ///
+    /// This replaces the previous three-pass approach:
```

Comment:
I don't think the comment here about old vs new approach is helpful / adds value

Could you also update this comment to clarify:
1. What is contained in the file_schema
2. What assumptions are made 

Specifically, I am confused as this code seems to assumes any columns referenced in `projection` that are greater than file_schema are partition columns. But what about columns that are in the table schema but not in the file schema (aka "schema evolution " columns that  will be filled in as nulls)

Reply:
I've added docstrings and updated the parameter name to be `logical_file_schema`.